core/src/main/scala/org/apache/spark/rdd/MapPartitionsRDD.scala

```diff
@@ -35,16 +35,24 @@ import org.apache.spark.{Partition, TaskContext}
  * @param isOrderSensitive whether or not the function is order-sensitive. If it's order
  *                         sensitive, it may return totally different result when the input order
  *                         is changed. Mostly stateful functions are order-sensitive.
+ * @param knownPartitioner If the result has a known partitioner.
  */
 private[spark] class MapPartitionsRDD[U: ClassTag, T: ClassTag](
     var prev: RDD[T],
     f: (TaskContext, Int, Iterator[T]) => Iterator[U],  // (TaskContext, partition index, iterator)
     preservesPartitioning: Boolean = false,
     isFromBarrier: Boolean = false,
-    isOrderSensitive: Boolean = false)
+    isOrderSensitive: Boolean = false,
+    knownPartitioner: Option[Partitioner] = None)
   extends RDD[U](prev) {
 
-  override val partitioner = if (preservesPartitioning) firstParent[T].partitioner else None
+  override val partitioner = {
+    if (preservesPartitioning) {
+      firstParent[T].partitioner
+    } else {
+      knownPartitioner
+    }
+  }
```

Comment:
We dont need to modify public api to add support for this.
Create a subclass of MapPartitionsRDD which has partitioner method overridden to specify what you need.
Did I miss something here ?

Reply:
`MapPartitionsRDD` is already private. But yes the other option is sub-classing.

Reply:
Since we are already creating a `MapPartitionsRDD` in distinct, overriding `partitioner` should be trivial.

Reply:
I mean yes we can sub-class just as easily -- is that what you mean?

Reply:
yes